### ISSUE TITLE
Improve behavior of "Select Ally 1-5" keybinds

### ISSUE THREAD

**Issue by Aluerie** [ValveSoftware/Dota2-Gameplay#9539]
I want to talk about the following key binds in the second page of Hotkeys settings: 
![image](https://github.com/ValveSoftware/Dota2-Gameplay/assets/33165440/34c24ff3-3387-4cc6-8f67-57b61c757606)

These keys were probably inspired by similar key binds from other MOBA games, however there is a lot of room for improvement. 

## Typical use case in other games: 

These keys in other games are mostly used for the following purpose: I want to check state of my allies, their map position, their exact wave state: I press F1, F2, F3, F4 - camera centers on each ally. So, with just 4 key presses I can cycle through my whole team and see what's everybody up to. Very convenient.

## How it's in Dota

To achieve the same I actually need to double tap these keys and, for some reason, we also cycle through ourselves, meaning in total we need to do F1, F1, F2, F2, F3, F3, F4, F4, F5, F5 - 10 key presses to achieve the same. 

## Suggestions

As we can see, what is achieved in other games in just 4 keypresses - requires 10 in Dota. This is why I propose the following changes:

- [ ] **Make the key bind center camera on selected ally by default.**
  Currently, it doesn't do that but just selects them, which is a bit pointless since I can view their items in the scoreboard and their HP/mana bars with Alt.

- [ ] **Do not cycle through ourselves**
  I only have four allies, which I want to cycle through, so why there is 5 key binds in the settings page. It would be nice if it was skipping ourselves in the cycle. If I'm teal then blue would be Ally 1 and Purple would be Ally 2.

**Comment by Aluerie** [comment]
Kinda similar suggestion to #2426 but they ask adjustable scoreboard and possibility to manually assign ally to each slot.

**Comment by Aluerie** [comment]
Unfortunately, multiple actions console binds do not work anymore in matchmaking so I can't even band-aid solve this by doing
```ps
bind F2 "dota_select_ally 2; dota_select_ally 2;"
```

I can script my f-keys so they double-tap themselves with some AutoHotKey scripts but that's borderline cheating so I'm hesitant to do that.

**Comment by commentator24** [comment]
are you preparing to use all 4 allies like meepo clones when they dc? 😅 
i hvnt try select ally..so x1 to select, x2 to centre camera is it?
i think it's confusing if purple can be 2 or 3. based on the order they appear in top bar is fine. with muscle memory its easier to remember purple is always 3.

its original use probably something like for healer to quickly check teammates hp + click HUD portrait to heal. it's a lot faster than using mouse to click top bar 1 by 1. hence if camera centre keep jumping around when you keep checking hp between 4 or 5 (including yourself), you might get dizzy headache soon... 😆 not all players want the camera to be centred.

If there are QOL improvements on this setting, I suggest new hotkeys to select enemy 1-5 as well. with the current ability to check out last known inventory and hp, i think it would be nice to be able to select enemy too. e.g. press x1 is like clicking 1x at top bar. press x2 centre camera to last known position (if fogged). this way teammates not confused where the enemy is.

Show inventory option in scoreboard should also show last known inventory (not current inventory) for both radiant and dire (instead of only allies), even better if can show last known courier items (not current), so a good captain who checked teammates items can plan ahead e.g. hey show inventory only reveal axe vanguard, but axe got vanguard and blink now in courier.. we should smoke gank soon.. i had games where teammates say bkb in courier but who even check courier items..?

if these are not possible, then I would recommend to be able to use select ally + phrases / chatwheel. e.g. you dont want to use voice chat and get reported so you press F3 (hold) then drag chatwheel: get back! or bait! or use alt ctrl shop item so it becomes i think someone (purple) should buy lotus then teammates understand you want only purple to getback / bait / buy lotus. this might be clearer than pinging teammates repeatedly or chatwheel get back / bait / and 3 player buy lotus.

Also chatwheel has Dive! but not Don't Dive! 😓 i would totally spam don't dive every game 👍

**Comment by Aluerie** [comment]
I'm never sure how to reply to your comments.

I'm also not sure if you played other Moba games, but the first link in [YouTube](https://youtu.be/i0KmQX2WbRw) shows how LoL pro player control the camera using such keys. I do not want to blabber but once you set up those keys to some convenient key-stroke suich as 2345 - it's really not that hard to spam once in a while and it brings a lot of useful info to the table.

> are you preparing to use all 4 allies

No, it's about checking information. 

> i think it's confusing if purple can be 2 or 3

then i can refer to suggestion n3. And well, it's my word vs yours. I think it's not confusing. In league there is no such problem :x 

> to quickly check teammate

You can just hold Alt to see that as I say in n1. 

The point is that exactly everything else except their detailed camera-centered map situation can be found with much shorter and more convenient ways than using Select Ally keys. The only original use to them is centering camera around allies thus it doesn't make sense for it not to be a default behavior.  

---

Eh well, it's whatever since I made a post at a garbage time (Friday evening) so nobody gonna upvote and Valve devs are on weekend break  to even notice it. So this post won't ever get implemented. Sadge 😢

**Comment by voddov** [comment]
I'm super agree with you, i was trying to set up some easy hotkeys for checking allies quickly but gave up just because the inconvenience and i belive there are probably way less than 1% of the players are actually using these hotkeys as it's currently poorly implemented.

**Comment by Glassman4443** [comment]
wow I was looking for a post like this, hoping it will be implemented one day, keeping the link in case it ever does.

Also, super disappointed to hear multiple action console binds doesn't work in actual games anymore,


I was so excited to test this out


alias "ToggleCameraFollow" "ToggleCameraFollowOn"
alias "ToggleCameraFollowOn" "+dota_camera_follow; alias ToggleCameraFollow ToggleCameraFollowOff; dota_camera_edgemove false"
alias "ToggleCameraFollowOff" "-dota_camera_follow; alias ToggleCameraFollow ToggleCameraFollowOn; dota_camera_edgemove true"
bind "\" "ToggleCameraFollow"

bind "F1" "+dota_camera_follow; ToggleCameraFollowOff"


Sadge.

**Comment by Glassman4443** [comment]
and yes, this is also a LoL concept, so many ideas from other games can be kept and used well.

**Comment by Aluerie** [comment]
I'm not sure what exactly you are trying to achieve with this wall of text binds but if I'm not mistaken your desire can be achieved just fine with 
```ps1
bindtoggle F1 "dota_camera_lock"
```

**Comment by j4n7** [comment]
Well. I miss that coming from LOL too. The thing is that without the possibility to rearrange the order of allies this is going to be useless.

This is such a good QoL.

**Comment by Glassman4443** [comment]
> ```powershell
> dota_camera_lock
> ```

Ye but then I would have to have another button that serves as an unlock no?

So in the end I went with "incrementvar dota_camera_lock 0 1 1", 

1 button to lock or unlock, 

dunno if dota camera lock could do it too, but in the end it worked out.

**Comment by Glassman4443** [comment]
> I'm not sure what exactly you are trying to achieve with this wall of text binds but if I'm not mistaken your desire can be achieved just fine with
> 
> ```powershell
> bindtoggle F1 "dota_camera_lock"
> ```
> 
> or maybe existing option "Hold Select Hero to Follow" is fine enough?

Ye but I wanted 1 button to lock or unlock, 

so in the end implemented it with "incrementvar dota_camera_lock 0 1 1"

**Comment by koala-lava** [comment]
It's done according to how you select your hero. But we want an option to just click it once. I have don't have a single button for that but need a combination so it makes even harder for me.  Also I want an option for that to act like "show last event button" which is to type it again to come back to your hero.

**Comment by Aluerie** [comment]
> This issue is stale because it has been open for 365 days with no activity.

Still valid.

Even though, I'm not sure if it's worth keeping this issue open with current "chaotic" structure of the repo. 

Our chances are probably better if we repost it once a year rather than keeping an open issue on 999th page. Idk.

**Comment by maakep** [comment]
I think you can do...

```
alias select0 "dota_camera_focus_player 0; dota_camera_focus_player 0"
alias select1 "dota_camera_focus_player 1; dota_camera_focus_player 1"

bind F1 select0
bind F2 select1
```
... etc.

If you just want one button to cycle through all allies, incrementvar could work:

``` 
bind F4 "incrementvar dota_camera_focus_player 0 4 1"
```

This binds F4 (or whatever hotkey you wish) to run the `dota_camera_focus_player X` command where X gets +1 every time you run it, within the span of 0 - 4 (incrementvar <command> <min> <max> <step>) 

> [!NOTE]
> This does not center the camera though

----

Because I usually want to focus on keeping tabs on just 1 specific hero, with 1 specific hotkey, here's what I do to account for the variability of that position:

```
alias cam1 "bind F4 dota_camera_focus_player 0; alias nextcam cam2"
alias cam2 "bind F4 dota_camera_focus_player 1; alias nextcam cam3"
alias cam3 "bind F4 dota_camera_focus_player 2; alias nextcam cam4"
alias cam4 "bind F4 dota_camera_focus_player 3; alias nextcam cam5"
alias cam5 "bind F4 dota_camera_focus_player 4; alias nextcam cam1"

alias nextcam cam1

bind F5 nextcam
```

> [!IMPORTANT]
> Just keep in mind that aliases needs to be added (afaik) to autoexec.cfg in your dota/cfg folder1.
> Windows example location: C:\Program Files (x86)\Steam\steamapps\common\dota 2 beta\game\dota\cfg

In short, this binds F5 to binding F4 to select a specific player. If I want F4 to select player 3, I click F5 3 times to make the "bind F4 dota_camera_focus_player 2" trigger. (I can also fairly easily cycle through many by spamming F5+F4, but that doesn't happen very much in practice. Usually you'll have some singular critical player you need to keep tabs on)

However, similarly, you could probably combine the logic here to create aliases the center the camera on a player, and bind that to one button that switches the "next player" so that you can go through all players, if that makes sense.

**Comment by Aluerie** [comment]
* multiple actions binds were disabled from public matchmaking since like 2015 so all binds with `;` won't work. I mention it in the 2nd comment of the thread.
* `dota_camera_focus_player` isn't a `cvar` - `incrementvar` won't work with it.

So nothing from the comment above will work.
